src/main/java/com/indeed/jiraactions/ActionsBuilder.java

```diff
@@ -43,78 +50,81 @@ public ActionsBuilder(final ActionFactory actionFactory, final Issue issue, fina
-                    actions.add(currentActionIndex + 1, commentAction);
-                    break;
-                } else {
-                    currentActionIndex++;
-                    if (currentActionIndex >= actions.size()) {
-                    /* You'd think this would never happen, but it can. I found legitimate examples with a comment
-                     * on a ticket *before* that ticket was created.
-                     */
-                        if (comment.created.isBefore(actions.get(0).getTimestamp())) {
-                            LOG.debug("Skipping comment {} on {} because it's before the issue was created.",
-                                    comment.id, issue.key);
-                        } else {
-                            LOG.debug("Unable to process comment {} by {} on issue {}, somehow doesn't fit in our timeline.",
-                                    comment.id, comment.author.getDisplayName(), issue.key, comment.author.getDisplayName());
-                        }
-                        currentActionIndex = 0;
+    //
+    // Updating Action to Given Date
+    //
+
+    private void setActionToCurrent() {
+        actions.set(0, actionFactory.toCurrent(actions.get(0)));
+    }
+
+    private void compare() {
+        int historyIndex = 0;
+        int commentIndex = 0;
+        for(Comment comment : comments) {
+            if (comment.isBefore(actions.get(0).getTimestamp())) {
+                LOG.debug("Skipping comment {} on {} because it's before the issue was created.", comment.id, issue.key);
+                commentIndex++;
+            }
+        }
+        while (true) {
+            if (historyIndex >= histories.size() || commentIndex >= comments.size()) {
+                if (commentIndex >= comments.size()) {
+                    if (historyIndex >= histories.size()) {
                         break;
+                    } else {
+                        setUpdateActions(historyIndex);
+                        historyIndex++;
                     }
+                } else {
+                    setCommentActions(commentIndex);
+                    commentIndex++;
+                }
+            } else {
+                if (histories.get(historyIndex).created.isBefore(comments.get(commentIndex).created)) {
+                    setUpdateActions(historyIndex);
+                    historyIndex++;
+                } else {
+                    setCommentActions(commentIndex);
+                    commentIndex++;
                 }
             }
         }
     }
 
-    private boolean isCreatedDuringRange(final DateTime createdDate) {
-        return startDate.compareTo(createdDate) <= 0 && endDate.compareTo(createdDate) > 0;
+    private List<History> sortLatestHistories(final Issue issue) {
+        issue.changelog.sortHistories();
+        final History[] histories = issue.changelog.histories;
+        final List<History> listHistories = new ArrayList<>(histories.length);
```

Comment:
Glad you preallocated the ArrayList size (although you could do new ArrayList<>(Arrays.asList(histories), I'm pretty sure).